File: da-test/src/test/java/ro/johann/da/test/api/service/decision/transfer/DecisionOutput.java
package ro.johann.da.test.api.service.decision.transfer;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.experimental.FieldDefaults;

import java.util.List;

import static com.fasterxml.jackson.annotation.JsonInclude.Include.NON_NULL;
import static lombok.AccessLevel.PRIVATE;

@Data
@Builder
@AllArgsConstructor
@NoArgsConstructor
@FieldDefaults(level = PRIVATE)
@JsonIgnoreProperties(ignoreUnknown = true)
@JsonInclude(NON_NULL)
public class DecisionOutput {
  String id;
  String name;
  String description;
  String status;
  AidSummary aidSummary;
  List<CriteriaOutput> criteria;
  List<AlternativeOutput> alternatives;
  List<PropertyOutput> properties;

  public void addAlternative(AlternativeOutput alternative) {
    alternatives.add(alternative);
  }

  public void addCriteria(CriteriaOutput criterion) {
    criteria.add(criterion);
  }
}

File: da-decision-service/src/main/kotlin/ro/johann/da/decision/api/transfer/DecisionOutput.kt
package ro.johann.da.decision.api.transfer

import com.fasterxml.jackson.annotation.JsonFormat
import ro.johann.da.decision.domain.Decision
import ro.johann.da.decision.domain.ProcessingResult
import java.time.LocalDateTime
import java.util.UUID

data class DecisionOutput(
  val id: UUID?,
  val name: String?,
  val description: String?,
  val status: String,
  val aidSummary: AidSummary? = null,
  val criteria: List<CriteriaOutput> = emptyList(),
  val alternatives: List<AlternativeOutput> = emptyList(),
  val properties: List<PropertyOutput> = emptyList(),
  @JsonFormat(pattern = "dd-MM-yyyy hh:mm:ss")
  val createdAt: LocalDateTime,
  @JsonFormat(pattern = "dd-MM-yyyy hh:mm:ss")
  val updatedAt: LocalDateTime
) {
  constructor(
    decision: Decision,
    processingResult: ProcessingResult? = null
  ) : this(
    id = decision.id,
    name = decision.name,
    description = decision.description,
    status = decision.status.value,
    aidSummary = processingResult?.let(::AidSummary),
    criteria = decision.criteria.map(::CriteriaOutput),
    alternatives = decision.alternatives
      .map { alternative -> AlternativeOutput(alternative, processingResult?.getProcessedAlternative(alternative.id)) },
    properties = decision.alternatives
      .flatMap { it.properties }
      .map { property -> PropertyOutput(property, processingResult?.getProcessedProperty(property.id)) },
    createdAt = decision.createdAt,
    updatedAt = decision.updatedAt
  )
}
